runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validator.go

```diff
@@ -853,3 +891,29 @@ func (e *AuthorizationMismatchError) Error() string {
 		e.FoundAuthorization.QualifiedString(),
 	)
 }
+
+// UnrepresentableEntitlementsUpgrade is reported during a contract upgrade,
+// when a composite or interface type is given access modifiers on its field that would
+// cause the migration to produce an unrepresentable entitlement set for references to that type
+type UnrepresentableEntitlementsUpgrade struct {
+	Type                 sema.Type
+	InvalidAuthorization sema.Access
+	ast.Range
+}
+
+var _ errors.UserError = &UnrepresentableEntitlementsUpgrade{}
+var _ errors.SecondaryError = &UnrepresentableEntitlementsUpgrade{}
+
+func (*UnrepresentableEntitlementsUpgrade) IsUserError() {}
+
+func (e *UnrepresentableEntitlementsUpgrade) Error() string {
+	return fmt.Sprintf(
+		"unsafe access modifiers on %s: the entitlements migration would grant references to this type %s authorization, which is too permissive.",
+		e.Type.QualifiedString(),
+		e.InvalidAuthorization.QualifiedString(),
+	)
+}
+
+func (e *UnrepresentableEntitlementsUpgrade) SecondaryError() string {
+	return "Consider removing any disjunction access modifiers"
```

Comment:
This definitely resolves the issue, but maybe this can be less "aggressive", e.g. leaving a single disjunction is also OK. Feel free to leave as-is for now.